keel-core/src/test/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentPromotionCheckerTests.kt

```diff
@@ -319,7 +357,129 @@ internal class EnvironmentPromotionCheckerTests : JUnit5Minutests {
 
         test("no artifact is registered") {
           verify(exactly = 0) {
-            artifactRepository.approveVersionFor(deliveryConfig, artifact, "1.0", environment.name)
+            repository.approveVersionFor(deliveryConfig, artifact, any(), environment.name)
+          }
+        }
+      }
+
+      context("the environment has a stateful constraint and a version cannot be found") {
+        deriveFixture {
+          copy(environment = Environment(
+            name = "staging",
+            constraints = setOf(ManualJudgementConstraint())
+          ))
+        }
+
+        before {
+          // TODO: sucks that this is necessary but when using deriveFixture you get a different mockk
+          every { repository.artifactVersions(artifact) } returns listOf("2.0", "1.2", "1.1")
+
+          every { repository.vetoedEnvironmentVersions(any()) } returns emptyList()
+
+          every { repository.pinnedEnvironments(any()) } returns emptyList()
+
+          every { repository.vetoedEnvironmentVersions(any()) } returns emptyList()
+
+          every { repository.pendingConstraintVersionsFor(any(), any()) } returns emptyList()
+
+          every { repository.getQueuedConstraintApprovals(any(), any()) } returns emptySet()
+
+          every { statefulEvaluator.canPromote(artifact, "2.0", deliveryConfig, environment) } returns false
+
+          every {
+            repository.constraintStateFor("my-manifest", "staging", "2.0")
+          } returns listOf(pendingManualJudgement)
+
+          runBlocking { subject.checkEnvironments(deliveryConfig) }
+        }
+
+        test("stateful constraints are only evaluated for the most recent version") {
```

Comment:
This covers the "adding a new environment (thus no approved versions) with a pipeline constraint attempts to launch a concurrent pipeline for every known artifact version" bug.